lib/storage/bucket.js

```diff
@@ -723,68 +734,255 @@ Bucket.prototype.setMetadata = function(metadata, callback) {
+  options = options || {};
+
+  var errors = [];
+  var files = [];
+
+  var globOptions = extend({}, options.globOptions, { nodir: true });
+
+  globby(pattern, globOptions, function(err, filePaths) {
+    if (err) {
+      callback(err);
+      return;
+    }
+
+    var uploadFileFns = filePaths.map(function(filePath) {
+      return function(next) {
+        var fileName;
+
+        if (options.basePath) {
+          fileName = path.relative(options.basePath, filePath);
+        } else {
+          fileName = path.basename(filePath);
+        }
+
+        var opts = extend({}, options, { destination: fileName });
+
+        self.uploadFile(filePath, opts, function(err, file) {
+          if (err) {
+            errors.push(err);
+          } else {
+            files.push(file);
+          }
+
+          next(options.force ? null : err);
+        });
+      };
+    });
+
+    async.parallelLimit(uploadFileFns, MAX_PARALLEL_UPLOADS, function() {
+      callback(errors, files);
+    });
+  });
+};
+
+/**
+ * Upload the contents of a directory to the root of your bucket. The structure
+ * of the given directory will be maintained.
+ *
+ * @param {string} directoryPath - Path to the directory you wish to upload.
+ * @param {object=} options - Configuration object.
+ * @param {string} options.basePath - A parent directory to use as the tip of
+ *     the resulting hierarchy in your bucket. File names are determined using
+ *     this value, which defaults to the given `directoryPath`. See the example
+ *     below for more.
+ * @param {boolean} options.resumable - Force a resumable upload. (default:
+ *     true for files larger than 5MB). Read more about resumable uploads
+ *     [here](http://goo.gl/1JWqCF). NOTE: This behavior is only possible with
+ *     this method, and not {module:storage/file#createWriteStream}. When
+ *     working with streams, the file format and size is unknown until it's
+ *     completely consumed. Because of this, it's best for you to be explicit
+ *     for what makes sense given your input.
+ * @param {string|boolean} options.validation - Possible values: `"md5"`,
+ *     `"crc32c"`, or `false`. By default, data integrity is validated with an
+ *     MD5 checksum for maximum reliability. CRC32c will provide better
+ *     performance with less reliability. You may also choose to skip validation
+ *     completely, however this is **not recommended**.
+ * @param {function} callback - The callback function.
+ *
+ * @example
+ * var zooPhotosPath = '/Users/stephen/Photos/zoo';
+ *
+ * bucket.uploadDirectory(zooPhotosPath, function(err, files) {
+ *   // Your bucket now contains:
+ *   //   - "monkeys/monkey-1.jpg"
+ *   //   - "zebras/zebra-1.jpg"
+ *   //   - "sleeping-panda.jpg"
+ * });
+ *
+ * //-
+ * // You can also specify a `basePath` if you need more control.
+ * //-
+ * var options = {
+ *   basePath: '/Users/stephen/Photos';
+ * };
+ *
+ * bucket.uploadDirectory(zooPhotosPath, options, function(err, files) {
```

Comment:
Methods that deal with multiple files don't accept destination, since that [historically] refers to a single file. So, uploadDirectory and upload just accept basePath, while uploadFile accepts destination.

At the end of the day, a user might want to handle all of this logic on their own, directly using uploadFile or even the bare-bones createWriteStream. It's not too late to scrap the PR and just let the user handle all this themselves :)